subprojects/core-api/src/main/java/org/gradle/api/plugins/ExtensionContainer.java

```diff
@@ -183,6 +183,81 @@
     @Nullable
     <T> T findByType(TypeOf<T> type);
 
+    /**
+     * Looks for the extension of a given type, or creates an adds a new extension to this container.
+     *
+     * A new instance of the given {@code type} will be created using the given {@code constructionArguments}.
+     * The extension will be exposed as {@code type} unless the extension itself declares a preferred public type via the {@link org.gradle.api.reflect.HasPublicType} protocol.
+     * The new instance will have been dynamically made {@link ExtensionAware}, which means that you can cast it to {@link ExtensionAware}.
+     *
+     * @param name The name for the extension
+     * @param type The type of the extension
+     * @param constructionArguments The arguments to be used to construct the extension instance
+     * @return The extension instance
+     * @throws IllegalArgumentException When an extension with the given name already exists.
+     * @see #findByType(Class)
+     * @see #create(String, Class, Object...)
+     * @since TODO
+     */
+    default <T> T findOrCreate(String name, Class<T> type, Object... constructionArguments) {
+        T extension = findByType(type);
```

Comment:
Here and in the following methods, it would probably be a good idea to verify the name also matches. Perhaps it would be more prudent to do (unchecked code, please verify locally):
```suggestion
        Object extensionByName = findByName(name);
        if (extensionByName != null && !type.isInstance(extensionByName)) {
            throw new IllegalStateException("Expected an extension of type " + type.getName() + " but found " + extensionByName.getClass().getName());
        }
        T extension = type.cast(extensionByName);
```